src/tracker/mime-type-tracker-query.vala

```diff
@@ -34,13 +38,40 @@ public class Games.MimeTypeTrackerQuery : Object, TrackerQuery {
 		return false;
 	}
 
-	public Game game_for_cursor (Tracker.Sparql.Cursor cursor) throws Error {
+	public void process_cursor (Tracker.Sparql.Cursor cursor) {
 		var uri = cursor.get_string (0);
+		uris += uri;
+	}
+
+	public async void foreach_game (GameCallback game_callback) {
+		uint handled_uris = 0;
+		foreach (var uri in uris) {
+			try {
+				var game = game_for_uri (uri);
+				game_callback (game);
+				handled_uris++;
+
+				// Free the execution only once every HANDLED_URIS_PER_CYCLE
+				// games to speed up the execution by avoiding too many context
+				// switching.
+				if (handled_uris >= HANDLED_URIS_PER_CYCLE) {
+					handled_uris = 0;
+
+					Idle.add (this.foreach_game.callback);
+					yield;
+				}
+			}
+			catch (Error e) {
+				debug (e.message);
+			}
+		}
+	}
 
+	public Game game_for_uri (string uri) throws Error {
```

Comment:
What about moving this to where it's called in `foreach_game` instead of renaming the callback?